Use `openspout` for Excel exports

chriscpty opened the pull request:
## Scope
This pull request includes a

- [x] Bug fix
- [x] New feature

## Changes
The following changes were made (this change is also documented in the [change log](https://github.com/kartik-v/yii2-export/blob/master/CHANGE.md)):

- (enh #357): *BC Breaking*: Use `openspout/openspout` rather than `PHPSpreadsheet` for XLSX and CSV exports
- (enh #378): Update `ExportWriterPdf::createExternalWriterInstance` declaration to be compatible with `PHPSpreadsheet` version 2

## Related Issues
If this is related to an existing ticket, include a link to it as well.

This is mainly about #357 , I also fixed #378 in the process.

chriscpty committed:
Use openspout for excel and CSV

+ use openspout for excel and CSV exports, rather than PHPSpreadsheet

chriscpty committed:
Use openspout for excel and CSV

+ various minor fixes

chriscpty committed:
Use openspout for excel and CSV

+ changelog

chriscpty committed:
Use openspout for excel and CSV

+ support more style options in OpenspoutHelper::createStyleFromPhpSpreadsheetOptions

chriscpty committed:
Use openspout for excel and CSV

+ autoWidth for export

chriscpty committed:
Use openspout for excel and CSV

+ adjusted autoWidth for export

chriscpty commented:
@kartik-v maybe instead of overriding the export type `FORMAT_EXCEL_X`, we should add a new type `FORMAT_EXCEL_X_FAST` that uses `openspout`? As it stands, `openspout` lacks some of the `PHPSpreadsheet` features the exporter uses, such as the `AutoFilter` function.

most of these we can hack solutions for (I already got a hacky autoSize to work and the box border styling should be doable too, albeit with some pain :D), but particularly with `AutoFilter` I don't really see how we'd be able to add that without going deep into `openspout`s internals.

Having `FORMAT_EXCEL_X_FAST` be a separate export type would allow users to choose performance or those additional features - plus the `onInit*` and `onRender*` hooks can more transparently not happen for that type specifically (rather than my current hack, where if any of those callbacks are set, we fall back to using `PHPSpreadsheet`)

What do you think?

kartik-v commented:
> @kartik-v maybe instead of overriding the export type `FORMAT_EXCEL_X`, we should add a new type `FORMAT_EXCEL_X_FAST` that uses `openspout`? As it stands, `openspout` lacks some of the `PHPSpreadsheet` features the exporter uses, such as the `AutoFilter` function.
> 
> most of these we can hack solutions for (I already got a hacky autoSize to work and the box border styling should be doable too, albeit with some pain :D), but particularly with `AutoFilter` I don't really see how we'd be able to add that without going deep into `openspout`s internals.
> 
> Having `FORMAT_EXCEL_X_FAST` be a separate export type would allow users to choose performance or those additional features - plus the `onInit*` and `onRender*` hooks can more transparently not happen for that type specifically (rather than my current hack, where if any of those callbacks are set, we fall back to using `PHPSpreadsheet`)
> 
> What do you think?

Yes I did mention in my comment on the other similar PR... that there must be an option to activate openspout selectively like a class property `useOpenSpout` ... I think the CSV, TEXT(Delimited) and XLSX formats can be easily derived from open spout . So it may not just be the XLSX format. For PDF output it may need formatting - need to put some effort to think about that later. Open spout also offers an additional output format *ODS*. So its better if we can include all natively supported formats by the library: i.e. when  `useOpenSpout` is `true` or then following formats can be set - else exception to be thrown:
- `FORMAT_EXCEL_X`
- `FORMAT_CSV`
- `FORMAT_TEXT`
- `FORMAT_ODS`

chriscpty commented:
I don't think it should throw an exception if `useOpenSpout` is true for a non-`openspout` format - instead, it should fall back to `PHPSpreadsheet`.
otherwise we run into a problem when you want a widget to support both` openspout`- and `PHPSpreadsheet`-exclusive formats - you'd have to set `useOpenSpout` in the export request to not throw an exception, rather than just setting it to `true` and leaving it there.

chriscpty committed:
Use openspout for excel and CSV

+ add useOpenspout option
+ add support for ODS exports
o rename _objExcel* objects to _objOpenspout* for clarity
! checks to prevent non-Excel openspout exports from crashing
o refactor out mergeCells function

chriscpty committed:
Use openspout for excel and CSV

+ add default options for ods export
! fix various minor issues
+ add empty i18n for openoffice

chriscpty committed:
Use openspout for excel and CSV

! fix ods export crashing if a cell has no border style
o minor improvements for ods export

chriscpty committed:
Use openspout for excel and CSV

+ complete i18n

chriscpty committed:
Use openspout for excel and CSV

+ add box styles for openspout export

chriscpty committed:
Use openspout for excel and CSV

+ add header styles back, refactor out outline/inside border logic

chriscpty committed:
Use openspout for excel and CSV

o last row check fix

chriscpty commented on src/ExportMenu.php:
```diff
@@ -1567,7 +1567,7 @@ public function generateBody()
             foreach ($models as $index => $model) {
                 $key = $keys[$index];
-                $isLastRow = $index === $totalCount;
+                $isLastRow = $index === $totalCount - 1;
```
One question about this specifically: Originally, this was just checking for `$index === $totalCount` - but when I tested this, that check never returned `true` as `$index` is zero-indexed.

Am I missing something here or was this an off-by-one-error before as well?

kartik-v replied:
> One question about this specifically: Originally, this was just checking for `$index === $totalCount` - but when I tested this, that check never returned `true` as `$index` is zero-indexed.
> 
> Am I missing something here or was this an off-by-one-error before as well?

Looks like a bug... in fact this condition needs to be made more robust and correct - even if we set to `totalCount - 1 ` it will fail if we had an `indexBy` clause applied on the query and the dataProvider will be indexed by some primary key instead of a sequential zero-based index. I think we need to run a counter inside the loop initialized from 1 and match that with totalCount as earlier (which was what it would have been and somehow maybe a PR had got it changed to a foreach index counter).

Something like below should be more correct - can you incorporate this update?

```php
$rowNum = 1;
foreach ($models as $index => $model) {
    $isLastRow = $rowNum === $totalCount;
    // other code
    $rowNum += 1;
}
```

chriscpty replied:
done now 🫡

chriscpty committed:
Use openspout for excel and CSV

o cleanup

chriscpty commented:
The good news: This is mostly working now and I think I've added all the functionality openspout supports to the openspout export.

The bad news: I don't think this actually performs better than PHPSpreadsheet? Maybe this is an implementation issue, I'll try and find out, but as of now the openspout export runs into memory issues at about the same point PHPSpreadsheet did in my testing :/

chriscpty committed:
Use openspout for excel and CSV

o fix german i18n
! fix bug with batching

chriscpty commented on src/ExportMenu.php:
```diff
@@ -1602,7 +1599,7 @@ public function generateBody()
                 $this->_provider->refresh();
                 $this->_provider->setTotalCount($totalCount);
-                $models = $this->_provider->getModels();
+                $models = array_values($this->_provider->getModels());
```
again, i'm pretty sure this was a preexisting bug - `array_values` is called on the first `getModels` call (see line 1547), but not called on later repetitions unless we are using an `ArrayDataProvider`. This crashed while I was testing with an `ActiveDataProvider` because `$index` would be >10, thus line 1566 would result in an out of bounds error.

kartik-v replied:
Yup this must be correct. Good catch.

chriscpty commented:
Nevermind the part about the performance not keeping up, i still had `YII_DEBUG` set 🤦 

This should be ready to review/merge now, I think

chriscpty committed:
Use openspout for excel and CSV

! better last row check

chriscpty committed:
Use openspout for excel and CSV

+ AutoFilter for openspout

chriscpty committed:
Use openspout for excel and CSV

o docs

chriscpty commented:
@kartik-v just to make sure this doesn't go under - I don't think I have anything further to add to this (I don't think we'll be able to add the currently missing features), so imo this is fully ready to review and merge

kartik-v commented:
> @kartik-v just to make sure this doesn't go under - I don't think I have anything further to add to this (I don't think we'll be able to add the currently missing features), so imo this is fully ready to review and merge

Thanks - before merging this I will just need to work a bit to release a minor version with other fixes. This PR will be merged with a higher major release. Plan to do this shortly.

ArtMin96 commented:
Any updates for this PR?

chriscpty commented:
One issue with this: as of current, I just pass a cell's `cellFormat` to openspout's `setFormat` function, but this leads to a disparity between openspout and PHPSpreadsheet - openspout expects a format like `0.00`, PHPSpreadsheet expects one of its `DataType::TYPE_*` constants.
I guess the way to go is to build an openspout format based on the PHPSpreadsheet format? although that would lose openspout's format customization options (which I need in this case, hence why I even came across this bug)

chriscpty commented:
i don't think this should be auto-closed, kartik has said he wants to merge this after he gets around to make a minor release
